### DISCUSSION TITLE
KeyError 'href' in python scraper with BeautifulSoup

### DISCUSSION THREAD

**Discussion by LuigiP99** [community/community#45566] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Show & Tell

### Body

Making a scraper script that extract links pagination from a amazon product page i've got an KeyError ''href'. And i know that the error is because when the scraper reach the last page, the next button is a span tag that contain the s-pagination-next class but also contain the s-pagination-disabled class.

I tried so many things to figure out how to solve it but nothing works. Any help would be appreciated

Here is the code:

```
import requests
from bs4 import BeautifulSoup


def getdata(url):
    headers = {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36",
        "Accept-Language": "en-US,en;q=0.5",
    }
    r = s.get(url, headers=headers).text
    soup = BeautifulSoup(r, "html.parser")
    return soup


def getnextpage(soup):
    next_ = soup.select_one("li.a-last a")
    if not next_:
        next_ = soup.select_one(".s-pagination-next")
        if not next_:
            return

    return "https://www.amazon.com" + next_["href"]


s = requests.Session()
url = f"https://www.amazon.com/s?k=sansumg.+galaxy+s22+ultra+desbloqueado&crid=2UIN9YKE55CMI&sprefix=%2Caps%2C1421&ref=nb_sb_ss_recent_1_0_recent"
while True:
    soup = getdata(url)
    # getdeals(soup)
    url = getnextpage(soup)
    if not url:
        break
    print(url)
```

**Comment by jge162** [comment, marked as answer]
@LuigiP99 

def getnextpage(soup):
    next_ = soup.select_one("li.a-last a")
    if not next_:
        next_ = soup.select_one(".s-pagination-next")
        if not next_:
            return
    if 'href' in next_.attrs:
        return "https://www.amazon.com" + next_["href"]


you could also check whether the class "s-pagination-disabled" is present on the span element. If present, it means the next button is disabled and there are no more pages to scrape. This can be done by adding a check for the class in the if statement.


   if 's-pagination-disabled' in next_['class']:
        return None


This should prevent the script from trying to access the href attribute of the next_ element when it is not present, and thus should resolve the KeyError: 'href' issue.

**Reply by LuigiP99** [reply]
I really appreciated your help, it works perfetly. I recently initialized into this and i've been practicing. 

Now i have another couple of problems, the script starts scraping from the next url of the current page, it starts count the second page as first page, i mean, the sr_pg_1 (the ending of the url that indicates the page number) is the sr_pg_2, and sr_pg_2 is the sr_pg_3, and so on. I can't scrape the url of the 'page one' because is the current page, the number one at the strip pagination is no enabled to scrape and not contains the url of the first page which would be the current page.

And the second problem is, in this case using this url, that have three pages, the scrape starts counting the first page as the second (the problem above). But the problem that is presented to me here is what the script finally returns and print in the console, some times it return just one of the two pages, and some times it returns the two pages (two of three pages but the first page is not counting)

I would really appreciate if you can help me with this.

**Reply by jge162** [reply]
No problem glad to have helped.

**Reply by jge162** [reply]
the first problem you described, where the script starts counting the second page as the first, would be to modify the script to start scraping from the current page instead of the next page

the second problem you mentioned, where the script sometimes returns only one or two of the pages, it could be that there is an issue with the logic that determines when to stop scraping. This could be caused by a bug in the script or by an issue with the website's pagination structure. One possible solution would be to add additional checks to the script to ensure that it is correctly identifying the last page and stopping the scraping when it reaches that point.
